frontend/src/metabase/entities/actions/utils.ts

```diff
@@ -0,0 +1,110 @@
+import _ from "underscore";
+
+import { getDefaultFieldSettings } from "metabase/actions/utils";
+
+import type {
+  ActionFormSettings,
+  FieldType,
+  InputSettingType,
+  Parameter,
+  ParameterType,
+} from "metabase-types/api";
+import type { TemplateTag, TemplateTagType } from "metabase-types/types/Query";
+import type NativeQuery from "metabase-lib/queries/NativeQuery";
+import type Question from "metabase-lib/Question";
+
+import {
+  fieldTypeToParameterTypeMap,
+  dateTypeToParameterTypeMap,
+  fieldTypeToTagTypeMap,
+} from "./constants";
+
+export const removeOrphanSettings = (
+  settings: ActionFormSettings,
+  parameters: Parameter[],
+): ActionFormSettings => {
+  const parameterIds = parameters.map(p => p.id);
+  const fieldIds = Object.keys(settings.fields);
+  const orphanIds = _.difference(fieldIds, parameterIds);
+
+  return {
+    ...settings,
+    fields: _.omit(settings.fields, orphanIds),
+  };
+};
+
+export const addMissingSettings = (
+  settings: ActionFormSettings,
+  parameters: Parameter[],
+): ActionFormSettings => {
+  const parameterIds = parameters.map(p => p.id);
+  const fieldIds = Object.keys(settings.fields);
+  const missingIds = _.difference(parameterIds, fieldIds);
+
+  if (!missingIds.length) {
```

Comment:
I personally don't like `!xxx.length` since it's not quite clear. When dealing with lengths I prefer we treat it as a number. That should be easier to read. What do you think?
```suggestion
  if (missingIds.length === 0) {
```